### PR TITLE
Improve bottom sheet handling of low available height

### PR DESCRIPTION
**Pull request by lcapkovic**
### Platforms Impacted
- [x] iOS
- [ ] macOS

### Description of changes

There can be cases where the height available to the bottom sheet is less than even the collapsed height. We're not handling this gracefully right now, letting the sheet go off screen and not dimming the content behind it.

The system sheet bypasses this issue by just going full screen whenever we're in compact height. Our sheet is more flexible and might not always have the full screen height available to it, so we can't rely on this metric (client could restrict our height arbitrarily even in regular screen height scenario).

To solve this:
- When the collapsed sheet height is sufficiently close to the max height, go into restricted height mode - the threshold for this is a constant and we might tweak it
- In this mode, collapsed state is effectively skipped and goes straight to expanded (if the sheet is expandable), because it doesn't make sense to have states that are just a few pts from each other.
- At the same time, ensure that both expanded and collapsed state never exceed the max available height.

While we're here - simplify our state animator to use the same dimming view / content alpha updating functions as everything else.

### Verification

Before

https://user-images.githubusercontent.com/3610850/163442271-0d7be647-fedd-43db-96da-24a60e4197e2.mov

After

https://user-images.githubusercontent.com/3610850/163442390-26023ea3-72c3-4610-966b-df771ec377d1.mov

### Pull request checklist

This PR has considered:
- [x] Light and Dark appearances
- [x] iOS supported versions (all major versions greater than or equal current target deployment version)
- [x] VoiceOver and Keyboard Accessibility
- [x] Internationalization and Right to Left layouts
- [x] Different resolutions (1x, 2x, 3x)
- [x] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)
- [x] iPad [Pointer interaction](https://developer.apple.com/documentation/uikit/pointer_interactions)
- [x] [SwiftUI](https://developer.apple.com/tutorials/swiftui) consumption (validation or new demo scenarios needed)
- [x] Objective-C exposure (provide it only if needed)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/fluentui-apple/pull/967)

### CODE REVIEW

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift [hunk 1]
```diff
@@ -858,6 +886,11 @@ public class BottomSheetController: UIViewController {
 
         static let maxSheetWidth: CGFloat = 610
         static let minSheetWidth: CGFloat = 300
+        static let defaultMaxSheetHeight: CGFloat = 600
+        
+        // When the difference in collapsed height and max sheet height is less than this,
+        // we go into height restricted mode which skips the collapsed state and adjusts dimming.
+        static let heightRestrictedThreshold: CGFloat = 50
```
Comment (huwilkes) [inline ios/FluentUI/Bottom Sheet/BottomSheetController.swift:893]: Where did 50 come from?

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift [hunk 2]
```diff
@@ -510,19 +508,37 @@ public class BottomSheetController: UIViewController {
...
+            // ----------------------------------------
+            // -------- fully undimmed (0.0)-----------
+            // ----------------------------------------
+            // ---------------------------------------- <--- high offset
+            
             let currentOffset = currentSheetVerticalOffset
-            let collapsedOffset = offset(for: .collapsed)
-            let expandedOffset = offset(for: .expanded)
+            let highestDimmedOffset = offset(for: .expanded)
+            let lowestUndimmedOffset = isHeightRestricted ? offset(for: .hidden) : offset(for: .collapsed)
```
Comment (huwilkes) [inline ios/FluentUI/Bottom Sheet/BottomSheetController.swift:532]: Would it be better to make the alpha independent of the offset when the height is restricted?

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift (hunk 1 above)
Comment (lcapkovic) [reply ios/FluentUI/Bottom Sheet/BottomSheetController.swift:893]: Magic number. The threshold has to be some number and this is what I came up with from testing of what distance feels sufficiently close to where a transition doesn't really make sense anymore.

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift (hunk 2 above)
Comment (lcapkovic) [reply ios/FluentUI/Bottom Sheet/BottomSheetController.swift:532]: This is in preparation for my next PR, which will allow swiping further down to go to hidden state. So we still need to accommodate interactive transition of alpha between .expanded and .hidden.

**Review by huwilkes — APPROVED** [review]

**Comment by huwilkes** [comment]
Looks like you have some whitespace lint error

**Comment by lcapkovic** [comment]
>

**Comment by lcapkovic** [comment]
> Looks like you have some whitespace lint error

Yeah... new machine, XCode not configured properly.

**Review by huwilkes — APPROVED** [review]

**Pull request merged**